runtime/major_gc.c

```diff
@@ -1566,5 +1566,8 @@ void caml_teardown_major_gc(void) {
 
 void caml_finalise_heap (void)
```

Comment:
The intention of `caml_finalise_heap` in OCaml 4 was that this would force the finalisers to run. In that spirit, freeing heap data structures here is incorrect. I would propose moving the functions `caml_teardown_shared_heap` and `caml_release_freelist` out of this function and place them after the call to `caml_finalise_heap` in `startup_aux.c` (or have them in a separate function which is called after `caml_finalise_heap`).